src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,223 @@
+        // This means any stores made to fields in the `value` object prior to this
+        // point cannot be reordered with the CAS operation of the reference to the
+        // `value` field.
+        // In other words, if a reader (using plain memory semantics) can observe a
+        // `value` reference, any field updates made prior to this fence are
+        // guaranteed to be seen.
+        UNSAFE.storeStoreFence(); // Redundant as a volatile put provides a store barrier?
+
+        // We are alone here under the `mutex`
+        // This upholds the invariant, the `@Stable value` field is written to
+        // at most once.
+        UNSAFE.putReferenceVolatile(this, VALUE_OFFSET, wrap(value));
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final T t = value();
+        if (t != null) {
+            return unwrap(t);
+        }
+        throw new NoSuchElementException("No value set");
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final T t = value();
+        if (t != null) {
+            return unwrap(t);
+        }
+        return other;
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return value() != null;
+    }
+
+    @ForceInline
+    @Override
+    public T computeIfUnset(Supplier<? extends T> supplier) {
+        final T t = value();
+        if (t != null) {
+            return unwrap(t);
+        }
+        return tryCompute(null, supplier);
+    }
+
+    @ForceInline
+    @Override
+    public <I> T computeIfUnset(I input, Function<? super I, ? extends T> function) {
+        final T t = value();
+        if (t != null) {
+            return unwrap(t);
+        }
+        return tryCompute(input, function);
+    }
+
+    @SuppressWarnings("unchecked")
+    @DontInline
+    private <I> T tryCompute(I input, Object provider) {
+        synchronized (mutex) {
+            // Updates to the `value` is always made under `mutex` synchronization
+            // meaning plain memory semantics is enough here.
+            T t = valuePlain();
+            if (t != null) {
+                return unwrap(t);
+            }
+            if (provider instanceof Supplier<?> supplier) {
+                t = (T) supplier.get();
+            } else {
+                t = ((Function<I, T>) provider).apply(input);
+            }
+            set0(t);
+            return t;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(orElse(null));
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return obj instanceof StableValueImpl<?> other &&
+                // Note that the value returned from `orElse()` can never be
+                // `nullSentinel()` (because it will be unwrapped to `null`).
+                // Therefore, we can safely use this as a marker for "no value set".
+                Objects.equals(orElse(nullSentinel()), other.orElse(nullSentinel()));
```

Comment:
This can use `StableValueImpl::value()` now:
```suggestion
                Objects.equals(value(), other.value());
```

Reply:
I personally am inclined to compare non-set stable values by their identity, so two non-set SVs should be distinct, given they can be set separately.

Reply:
The `equals()/hascode()` principles proposed here align with `Optional`.